ChangeLog

```diff
@@ -68,6 +68,11 @@ Release date: TBA
 
   Closes #4681
 
+* Fix false negative for ``used-before-assignment`` when the variable is assigned
+  in an exception handler, but used outside of the handler.
+
+  Closes #626
```

Comment:
I think it also closes #4391?

Reply:
Unfortunately it doesn't---the changes don't handle shadowing:

```python
def function():
    try:
        1 / 0
    except ZeroDivisionError as error:
        try:
            1 / 0
        except ZeroDivisionError as error:
            raise Exception("") from error
```

You can also see that there are tests in `tests/functional/u/unused/unused_variable.py:108-158` that still have TODOs in them.

Both occurrences of the `as error` are part of the same consumer in "`to_consume`", and they're both removed when the `error` `astroid.Name` node in the inner block is visited. I can think of a couple of different ways of fixing this, but none that are particularly easy/elegant. I can point out that we get a similar issue with loop variable shadowing:

```python
def function():
    for i in range(10):
        for i in range(20):
            print(i)
```

Pylint currently reports a `redefined-outer-name` error for the inner `i` loop variable, but does *not* report `unused-variable` for the outer `i` loop variable. The `redefined-outer-name` check is handled in a separate `visit_for` method; I could do something similar in `visit_excepthandler` if you want.

P.S. I'll reply to your general review comment later, need to grab lunch now!

Reply:
Thank you for your very clear answer !